Lb302 cleanup

rubiefawn opened the pull request:
Makes the following changes to the Lb302 plugin for increased maintainability and readability:
- [x] Use [non-static data member initializers](https://en.cppreference.com/w/cpp/language/data_members.html#Member_initialization) where possible
- [x] Use `std::numbers::pi` instead of `M_PI`
- [x] Use `std::array` and `std::unique_ptr` instead of manual memory management
- [x] Change plain-old-data `class`es with only public members to `struct`s
- [x] Change some `int`s to `f_cnt_t`, ~~`fpp_t`~~, etc. to better communicate their purpose and avoid weird implicit casting nonsense
- [x] Change `float`s to `sample_t` where appropriate to better communicate their purpose
- [x] Find suitable homes for loose constants
- [x] Prefix standard library math function calls with `std::`
- [x] Ensure all floating-point literals intended to be `float`s end in `f` and are not `double`s to appease MSVC
- [x] Change uses of `SIGNAL` and `SLOT` macros to whatever they should be post-Qt6 upgrade
- [x] Remove cursed mutex in favor of a real-time safe queue (TODO: possible generalization for #6516?)
- [ ] Optimize `struct`/`class` member layout for better locality
- [x] Completely conform to current code conventions

The above list is not comprehensive and is destined to expand.

Also adds note velocity and note panning to Lb302, as well as floating point denormal flush-to-zero on ARM to audio worker threads because might as well

# Testing checklist

- [ ] General Lb302 torture test: slam it with notes, arpeggios, chords, rapid automation of the slide parameter, anything you can think of. It's monophonic so it should only play one note at a time. Mainly looking for any performance or stability issues.
- [ ] Velocity behavior, especially with slide enabled
- [ ] Note panning behavior, especially with slide enabled
- [x] Every Lb302 preset we ship still loads and plays, and sounds no different than they did before

rubiefawn committed:
Lb302 cleanup part 1

Makes the following changes to the Lb302 plugin:
- Use non-static data member initializers where possible
- Find suitable homes for several loose constants
- Use std::numbers::pi instead of M_PI
- Use std::array and std::unique_ptr instead of manual memory management
- Change some plain-old-data classes with only public members to structs
- Change some ints to more descriptive types such as f_cnt_t, fpp_t, etc
- Attempt to conform any lines I touched to current code conventions

rubiefawn committed:
Combine and clean up `extern "C"` blocks

rubiefawn committed:
Prefix `tanh()` with `std::`

This should be the last of the functions from `<cmath>`

rubiefawn committed:
Change `float`s that are audio to `sample_t`

rubiefawn committed:
Modernize Qt `connect()` invocations

Begone, SIGNAL() SLOT() macro magic!!

rubiefawn committed:
Change remaining macro constants to constexpr vars

...and move them to be inside the classes that own them

rubiefawn committed:
Lb302 cleanup part 2

- Remove Lb302Note as it is just function parameters with extra steps
- Clean up Lb302Synth::initNote()
- Clean up Lb302Synth::process()
- Move the GET_INC() helper to an anon namespace, rename to phaseInc()
- Make computeDecayFactor() a lambda next to its only callsite

rubiefawn committed:
Change `double` literals to `float` literals

To avoid implicit casts, it's all supposed to be floats

rubiefawn committed:
Start to rearrange the layout of Lb302Synth

rubiefawn committed:
Code conventions tomfoolery

rubiefawn committed:
Make some more things constexpr

rubiefawn committed:
Minor coding conventions conformity

rubiefawn committed:
Code conventions applied to `Lb302SynthView`

rubiefawn committed:
Replace mutex with realtime-safe mpsc channel

rubiefawn commented:
Figured I'd run some un-rigorous `perf stat` benchmarks of release builds as a sanity check to make sure I wasn't destroying performance by changing the mutex for a bunch of atomic operations and busy-wait loops.

Updated for commit 03cca1bdb1d0332f5a66186b53c8e78c58117a9c

Before, with mutex/dynamically allocated vector:

```
Performance counter stats for './lmms render /home/fawn/Documents/lmms/projects/torture.mmpz -a -f flac':

   103,084,983,781      task-clock                       #    2.979 CPUs utilized             
           695,800      context-switches                 #    6.750 K/sec                     
               388      cpu-migrations                   #    3.764 /sec                      
             6,173      page-faults                      #   59.883 /sec                      
   267,717,430,944      instructions                     #    1.17  insn per cycle            
   229,289,575,609      cycles                           #    2.224 GHz                       
    61,653,273,269      branches                         #  598.082 M/sec                     
       496,591,833      branch-misses                    #    0.81% of all branches           

      34.602878156 seconds time elapsed

      93.334079000 seconds user
       9.660192000 seconds sys
```

After, with atomic ringbuffer/statically allocated array:

```
Performance counter stats for './lmms render /home/fawn/Documents/lmms/projects/torture.mmpz -a -f flac':

   104,963,456,247      task-clock                       #    3.066 CPUs utilized             
           708,920      context-switches                 #    6.754 K/sec                     
                67      cpu-migrations                   #    0.638 /sec                      
             4,237      page-faults                      #   40.366 /sec                      
   267,503,252,259      instructions                     #    1.15  insn per cycle            
   233,160,958,544      cycles                           #    2.221 GHz                       
    61,493,266,240      branches                         #  585.854 M/sec                     
       511,104,882      branch-misses                    #    0.83% of all branches           

      34.239441489 seconds time elapsed

      95.604870000 seconds user
       9.199340000 seconds sys
```

Project file used for said benchmarks: [torture.mmp.xml](https://github.com/user-attachments/files/24028033/torture.mmp.xml)

**Edit:** Turns out the arpeggiator has significant overhead, so these results aren't really meaningful.

rubiefawn committed:
Add missing headers

Also don't bother checking for __ARM_ACLE

rubiefawn committed:
Fallback hardware_destructive_interference_size

rubiefawn committed:
GCC is missing some ARM intrinsics, attempt asm

whyyyyyyy

rubiefawn committed:
Remove mutex header include

rubiefawn committed:
Process new notes last (previous behavior)

Also don't drop notes when the queue is full, busy-wait a limited
number of iterations to see if space gets freed up

rubiefawn committed:
Add documentation for all the ringbuffer stuff

Rossmaxx commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -28,92 +28,98 @@
 #define LB_24_IGNORE_ENVELOPE
-#define LB_FILTERED
-//#define LB_DECAY
 //#define LB_24_RES_TRICK
```
Since we cleaning up the codebase, it would make sense to remove this commented macro.

messmerd commented:
> `CACHELINE` and `busy_wait_hint` probably don't belong inside Lb302. Where is the best place for me to move these to?

Maybe add them to a new `LmmsPolyfill.h` header. And rename `CACHELINE` to `lmms::hardware_destructive_interference_size` to match the name the standard library uses.

rubiefawn committed:
Move hardware-specific stuff to Hardware.h

- Rename CACHELINE to lmms::hardware_destructive_interference_size
- Move lmms::hardware_destructive_interference_size and busy_wait_hint() to Hardware.h

rubiefawn committed:
Inline some functions only used in one place

messmerd commented on plugins/Lb302/Lb302.h:
```diff
@@ -71,109 +69,111 @@ class Lb302FilterKnobState
+	float vcf_c0 = 0.f; // c0=e1 on retrigger; c0*=ed every sample; cutoff=e0+c0
+	float vcf_e0 = 0.f; // e0 and e1 for interpolation
+	float vcf_e1 = 0.f;
+	float vcf_rescoeff; //!< Resonance coefficient [0.30, 9.54]
```
These could be renamed to follow the naming conventions for data members

rubiefawn replied:
Resolved as of a4769edc782473c690631477b0310f0fc2f61ef2

messmerd commented on plugins/Lb302/Lb302.h:
```diff
@@ -71,109 +69,111 @@ class Lb302FilterKnobState
-	float vcf_d1,           //   d1 and d2 are added back into the sample with
-	      vcf_d2;           //   vcf_a and b as coefficients. IIR2 resonance
-	                        //   loop.
+protected:
```
The data members of this class could also be renamed to follow the conventions

rubiefawn replied:
Resolved as of a4769edc782473c690631477b0310f0fc2f61ef2

rubiefawn committed:
Fixes for Hardware.h

- #include <new> *before* testing macro defined within it (lol)
- Make #ifdef chain for hardware_destructive_interference_size more readable
- Use LMMS_HOST_* macros from lmmsconfig.h where possible
- Move #includes for busy_wait_hint() to top of file
- Make busy_wait_hint() a function rather than a macro
- Minor formatting changes

rubiefawn commented on include/Hardware.h:
```diff
@@ -0,0 +1,72 @@
+#endif
+
+
+#if defined(__x86_64__) || defined(_M_AMD64) || defined(__i386__) || defined(_M_IX86)
```
Mostly resolved as of [`6fccf95`](https://github.com/LMMS/lmms/pull/8132/commits/6fccf9562d3f9fa5b5b557c6ae0795476594ad5b) (MSVC is a pain in the ass as usual)

rubiefawn committed:
Ignore parameter for __isb() polyfill

15 is the only allowed value, so that is used regardless of whatever is
passed in

rubiefawn committed:
Change static member names to fit conventions

rubiefawn committed:
Remove weird float pass-by-reference

rubiefawn committed:
Oops, missed one static member rename

rubiefawn committed:
Move phaseInc() helper into lmms namespace

rubiefawn committed:
Fix bug introduced in 1b563eda28f38aa914484bf29f3fb6f970ddc9d1

- Restored a line that resets the slide-from note when a slide is done,
  the accidental removal of which was the cause of a weird bug
- Actually remove initSlide() and initNote() which were supposed to be
  fully inlined

rubiefawn committed:
Resolve two TODOs

- Remove redundant `db24Toggled()` calls on init and preset load
- Move filter envelope decay stuff that does not require `recalcFilter()`
  from `filterChanged()` into new slot `decayChanged()`

rubiefawn committed:
Add & update documentation and comments

Unrelated to this commit, but I just realized there's an ancient README.
That may be worth reviewing and either updating or deleting it,
depending on its contents.

rubiefawn committed:
Stick some filter members in a POD struct

rubiefawn committed:
Move release_frame calculation to processNote()

rubiefawn committed:
Remove obsolete amplitude envelope hacks

Also update/fix some comments

sakertooth commented:
>  I'm still iffy on the idea of changing program-wide function signatures to accommodate one plugin, but there might be a way to avoid that too.

Yeah I would recommend not doing anything like this here (might not work and also blurs the responsibilities of this PR), but its something to think about.

We might consider also removing `InstrumentPlayHandle` in the future (among some of the others since they don't do any actual processing on their own, like what is `PresetPreviewPlayHandle` even doing here?) after figuring out a more correct design, so the proposed idea may not be applicable.

rubiefawn committed:
Change Lb302Synth member names to fit conventions

Also remove some old commented-out code and TODOs that were no longer
relevant

rubiefawn committed:
Remove Lb302Filter3Pole::kp1

It was only referenced in one function, so I made it a local variable

rubiefawn committed:
Change Lb302Filter3Pole member names to fit conv.

Also combine filter stage values into an array, use sample_t over float

rubiefawn committed:
Add note velocity to Lb302

rubiefawn committed:
Add note panning to Lb302

rubiefawn committed:
Style fixes

rubiefawn commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +155,622 @@ float Lb302FilterIIR2::process(const float& samp)
-	aout    = kp1h * (ay2 + ay31) - kp*aout;
-
-	return tanh(aout*value)*LB_24_VOL_ADJUST/(1.0+fs->dist);
+	static constexpr float s_volAdjust = 3.f;
```
Fixed in 43c3b3855c4b976fc47ae3b024a155eadedd7040

rubiefawn commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +155,622 @@ float Lb302FilterIIR2::process(const float& samp)
 	{
-		vca_mode = VcaMode::Decay;
+		constexpr float s_volRatio = 1.f / DefaultVolume;
+		m_noteVolume = m_playingNote->getVolume() * s_volRatio;
```
Fixed in 43c3b3855c4b976fc47ae3b024a155eadedd7040

rubiefawn committed:
Move contents of denormals.h to Hardware.h

Also use busyWaitHint() instead of raw _mm_pause() call

rubiefawn committed:
Attempt to add ARM FTZ to disableDenormals()

I don't have an ARM machine so time for the CI to tell me if this works.
I suspect GCC might be stupid with use of ACLE. Hopefully not, since I'd
prefer to use ACLE rather than inline assembly

rubiefawn committed:
Evil hacks

rubiefawn committed:
Does the register name need to be uppercase?

rubiefawn committed:
Give up on ACLE and use inline asm

despair

rubiefawn committed:
Merge branch 'master' into refac/lb302

rubiefawn committed:
Fix ordering of oscillator shapes in enum

rubiefawn committed:
Fix segfault on Windows

rubiefawn committed:
Add my name to Lb302 copyright headers

Since significant changes have been made lol

rubiefawn committed:
Minor formatting changes

rubiefawn committed:
Calculate note volume & panning once per buffer

...instead of once per sample. This mistake single-handedly added 5
seconds of render time to my Lb302 torture test. Wtf was I thinking lmao

rubiefawn commented:
Did some more benchmarks using `perf stat`, this time with a project file that doesn't use the arpeggiator: [lb302-torture.mmp.xml](https://github.com/user-attachments/files/26336356/lb302-torture.mmp.xml)

# Current master (098f0ed4b99b09d2a1707b5b6002e486abb092f0)

**Edit:** I was previously using the AUR `lmms-git` package, but that wasn't compiled with the same CMake flags as the other test. I re-ran the test after compiling `master` locally with the same CMake flags for a more accurate comparison.

```
Performance counter stats for '~/src/lmms/build/lmms render lb302-torture.mmp -a -f flac' (10 runs):

430,971,240,352 instructions     ( +-  0.13% )
811,411,815,600 cycles           ( +-  0.40% )
     25,977,738 context-switches ( +-  0.28% )
         40,556 page-faults      ( +-  0.01% )
  4,969,657,842 branch-misses    ( +-  0.27% )
  7,690,457,618 cache-misses     ( +-  0.61% )

49.331 +- 0.148 seconds time elapsed  ( +-  0.30% )
```

# This branch (c86ef78434fb2e7360f5941fb67b36ded4e4f552)

```
Performance counter stats for '~/src/lmms-lb302/build/lmms render lb302-torture.mmp -a -f flac' (10 runs):

400,493,308,746 instructions     ( +-  0.07% )
758,790,284,354 cycles           ( +-  0.28% )
     24,368,422 context-switches ( +-  0.15% )
         40,361 page-faults      ( +-  0.21% )
  4,765,039,651 branch-misses    ( +-  0.17% )
  8,009,820,790 cache-misses     ( +-  0.67% )

44.6630 +- 0.0568 seconds time elapsed  ( +-  0.13% )
```

rubiefawn commented on src/core/AudioEngineWorkerThread.cpp:
```diff
@@ -99,12 +96,7 @@ void AudioEngineWorkerThread::JobQueue::run()
-		_mm_pause();
-#endif
-	}
+	while (m_itemsDone < m_writeIndex) { busyWaitHint(); }
```
I shouldn't worry about it in this PR, but these atomic member variables should probably use explicit memory ordering everywhere

sakertooth commented:
> Did some more benchmarks using `perf stat`, this time with a project file that doesn't use the arpeggiator: [lb302-torture.mmp.xml](https://github.com/user-attachments/files/26336356/lb302-torture.mmp.xml)
> 
> 
> 
> # Current master (098f0ed4b99b09d2a1707b5b6002e486abb092f0)
> 
> 
> 
> **Edit:** I was previously using the AUR `lmms-git` package, but that wasn't compiled with the same CMake flags as the other test. I re-ran the test after compiling `master` locally with the same CMake flags for a more accurate comparison.
> 
> 
> 
> ```
> 
> Performance counter stats for '~/src/lmms/build/lmms render lb302-torture.mmp -a -f flac' (10 runs):
> 
> 
> 
> 430,971,240,352 instructions     ( +-  0.13% )
> 
> 811,411,815,600 cycles           ( +-  0.40% )
> 
>      25,977,738 context-switches ( +-  0.28% )
> 
>          40,556 page-faults      ( +-  0.01% )
> 
>   4,969,657,842 branch-misses    ( +-  0.27% )
> 
>   7,690,457,618 cache-misses     ( +-  0.61% )
> 
> 
> 
> 49.331 +- 0.148 seconds time elapsed  ( +-  0.30% )
> 
> ```
> 
> 
> 
> # This branch (c86ef78434fb2e7360f5941fb67b36ded4e4f552)
> 
> 
> 
> ```
> 
> Performance counter stats for '~/src/lmms-lb302/build/lmms render lb302-torture.mmp -a -f flac' (10 runs):
> 
> 
> 
> 400,493,308,746 instructions     ( +-  0.07% )
> 
> 758,790,284,354 cycles           ( +-  0.28% )
> 
>      24,368,422 context-switches ( +-  0.15% )
> 
>          40,361 page-faults      ( +-  0.21% )
> 
>   4,765,039,651 branch-misses    ( +-  0.17% )
> 
>   8,009,820,790 cache-misses     ( +-  0.67% )
> 
> 
> 
> 44.6630 +- 0.0568 seconds time elapsed  ( +-  0.13% )
> 
> ```

I would recommend scoping `perf` to only the audio threads to avoid skewing the results with the main thread and the like.

rubiefawn committed:
Merge branch 'master' into refac/lb302

bratpeki commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -28,165 +28,121 @@
 
-#include "Lb302.h"
 #include "AutomatableButton.h"
+#include "BandLimitedWave.h"
```
Why is this included now?

rubiefawn replied:
It was always included, I just organized the includes according to the code style guidelines. If you look a little further down you'll see it was removed from what used to be line 46.

bratpeki commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -28,165 +28,121 @@
-//#define engine::audioEngine()->outputSampleRate() 44100.0f
-const float sampleRateCutoff = 44100.0f;
+// Helper to get the phase increment per sample, given a note's frequency and the current sample rate
+static inline float phaseInc(float freq) { return freq / Engine::audioEngine()->outputSampleRate(); }
```
Note to self: Replaces `GET_INC`.

bratpeki commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,621 @@ float Lb302FilterIIR2::process(const float& samp)
 }
 
 
+void Lb302Synth::db24Toggled() { recalcFilter(); }
```
Why was `vcf = vcfs[db24Toggle.value()];` removed?

rubiefawn replied:
Because I removed the `vcf` member entirely and replaced it with a getter function of the same name, no need to store redundant state. It's used on line 313 and 330.

bratpeki commented on plugins/Lb302/Lb302.h:
```diff
@@ -71,191 +66,246 @@ class Lb302FilterKnobState
+	void recalcFilter();
 
-	FloatModel vco_fine_detune_knob;
+	enum class VcoShape { Sawtooth, Triangle, Square, RoundSquare, Moog, Sine, Exponential, WhiteNoise,
```
The old one was:
```cpp
enum class VcoShape { Sawtooth, Square, Triangle, Moog, RoundSquare, Sine, Exponential, WhiteNoise, BLSawtooth, BLSquare, BLTriangle, BLMoog };
```
Does the order matter for the models? Cause it's not being preserved.

rubiefawn replied:
The order change is intended; see https://github.com/LMMS/lmms/pull/8132#issuecomment-4132370201

bratpeki commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,621 @@ float Lb302FilterIIR2::process(const float& samp)
+	const auto readIdx = m_notesReadSeq.load(std::memory_order_relaxed);
+	const auto writeCommitted = m_notesWriteCommitted.load(std::memory_order_acquire);
+	// Process notes, but process new notes last
+	for (size_t i = readIdx; i < writeCommitted; ++i)
+	{
+		const auto& nph = m_notes[i & s_notesBufMask];
+		if (nph->totalFramesPlayed() == 0) { continue; }
+		processNote(nph);
+	}
+	for (size_t i = readIdx; i < writeCommitted; ++i)
+	{
+		const auto& nph = m_notes[i & s_notesBufMask];
+		if (nph->totalFramesPlayed() != 0) { continue; }
+		processNote(nph);
+	}
```
Why does this happen twice?

rubiefawn replied:
The first loop processes old notes, the second processes new notes. (Or rather, the first loop skips new notes, while the second loop skips old notes.)

bratpeki replied:
How exactly does that work? They look identical.

rubiefawn replied:
They are identical, except for the `continue` condition: `==` vs `!=`.

rubiefawn committed:
Add 32-bit x86 and ARM support to Hardware.h

rubiefawn committed:
Add RISC-V support to lmms::busyWaitHint()

Technically it's an extension ("Zihintpause"), but it is encoded as a
FENCE instruction so it should be ineffective but harmless on
implementations without the extension

messmerd commented on plugins/Lb302/CMakeLists.txt:
```diff
@@ -1,3 +1,8 @@
+IF(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+	ADD_COMPILE_OPTIONS(-Wno-interference-size)
+ENDIF()
+
```
Maybe this could be moved to `BuildPlugin.cmake` so it applies to all plugins?

rubiefawn replied:
I like this idea, but I get about ten thousand of the following error, mostly from LADSPA plugins with C sources:

```
warning: command-line option ‘-Wno-interference-size’ is valid for C++/ObjC++ but not for C
```

**Edit:** I wonder if I can put the solution inside `Hardware.h` itself by using `#pragma GCC diagnostic ignored`?

**Edit 2:** Resolved as of 19023f4e9583f6c0455db3c8bcb8d6ce6e065b25.

messmerd commented on include/Hardware.h:
```diff
@@ -0,0 +1,144 @@
+	// Intel® 64 and IA-32 Architectures Software Developer's Manual Volume 1: Basic Architecture
+	// 11.6.3 Checking for the DAZ Flag in the MXCSR Register
+	unsigned int flags = 0x8000; // FTZ
+	std::uint8_t buf[512] = {0};
```
```suggestion
	alignas(16) std::uint8_t buf[512] = {0};
```
According to [this](https://www.felixcloutier.com/x86/fxsave), this has to be aligned to a 16-byte boundary.

rubiefawn replied:
Interesting. I didn't see anything about alignment on the *Intel Intrinsic Guide* for [`_fxsave(), _fxsave64()`](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=fxsave&ig_expand=3523,3522) so I assumed it wasn't necessary. The *Intel® 64 and IA-32 Architectures Software Developer’s Manual* **does** specify that alignment requirement, but I didn't even think to check that since the *Intel Intrinsic Guide* usually includes alignment requirements.

I'll go ahead and add that back.

**Edit:** Resolved as of 19023f4e9583f6c0455db3c8bcb8d6ce6e065b25.

messmerd commented on plugins/Lb302/Lb302.h:
```diff
@@ -71,191 +66,246 @@ class Lb302FilterKnobState
+	//! @brief Helper to get current vcf
+	//! @see m_vcfs
+	//! @see db24Toggle
+	inline Lb302Filter& vcf() { return *m_vcfs[m_db24Toggle.value()]; }
```
```suggestion
	Lb302Filter& vcf() { return *m_vcfs[m_db24Toggle.value()]; }
```
This method is already implicitly inline

rubiefawn replied:
Resolved as of 73b92f4f532b016c81fecc6dd4f5731d201d0520

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,621 @@ float Lb302FilterIIR2::process(const float& samp)
+	auto tries = s_maxNoteEnqueueRetries;
+	auto write_claimed_expected = m_notesWriteClaimed.load(std::memory_order_relaxed);
+	size_t index, next_index;
+	do
```
I think you could remove the `goto` with a little reorganizing:

```cpp
for (;;)
{
	if (!tries--)
	{
		qDebug() << "Lb302: Note dropped due to catastrophically poor performance! This should never happen!";
		return;
	}

	const std::size_t occupied = writeClaimedExpected - m_notesReadSeq.load(std::memory_order_acquire);
	assert(occupied >= 0);

	// If full, wait for room
	if (occupied >= s_maxPendingNotes)
	{
		busyWaitHint();
		writeClaimedExpected = m_notesWriteClaimed.load(std::memory_order_relaxed);

		// Try to send note again
		continue;
	}

	index = writeClaimedExpected;
	nextIndex = writeClaimedExpected + 1;

	if (m_notesWriteClaimed.compare_exchange_strong(writeClaimedExpected, nextIndex, std::memory_order_acquire))
	{
		// Successfully sent note
		break;
	}
}
```

I wasn't sure why `occupied` was a `ptrdiff_t` - I made it a `std::size_t` instead.

rubiefawn replied:
> I wasn't sure why occupied was a `ptrdiff_t` - I made it a `std::size_t` instead.

This is because I wanted to catch the case for `a - b` where `a < b`, hence the `assert()` on the next line. `ptrdiff_t` is signed, which allows me to distinguish between the case when `occupied < 0` (a fatal error) vs `occupied >= s_maxPendingNotes` (expected when the queue is full).

However, I just realized it wouldn't work as intended since it's currently just going to underflow to a really big number first (both operands are `size_t`), then put that really big number into the signed type lol. I need to cast the operands to a signed type *first* before subtracting.

rubiefawn replied:
Resolved as of 19023f4e9583f6c0455db3c8bcb8d6ce6e065b25

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,621 @@ float Lb302FilterIIR2::process(const float& samp)
-		//vca_a = 0.0;
+	auto tries = s_maxNoteEnqueueRetries;
+	auto write_claimed_expected = m_notesWriteClaimed.load(std::memory_order_relaxed);
+	size_t index, next_index;
```
```suggestion
	auto writeClaimedExpected = m_notesWriteClaimed.load(std::memory_order_relaxed);
	std::size_t index;
	std::size_t nextIndex;
```
Fixed naming convention, used `std::`, and put declarations on their own lines

rubiefawn replied:
Resolved as of 19023f4e9583f6c0455db3c8bcb8d6ce6e065b25

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,621 @@ float Lb302FilterIIR2::process(const float& samp)
+	size_t write_committed_expected = index;
+	while (!m_notesWriteCommitted.compare_exchange_strong(write_committed_expected, next_index, std::memory_order_release))
+	{
+		write_committed_expected = index; // Reset this as the CAS will have changed it
```
```suggestion
	std::size_t writeCommittedExpected = index;
	while (!m_notesWriteCommitted.compare_exchange_strong(writeCommittedExpected, nextIndex, std::memory_order_release))
	{
		writeCommittedExpected = index; // Reset this as the CAS will have changed it
```
Naming convention and using `std::`

rubiefawn replied:
Resolved as of 19023f4e9583f6c0455db3c8bcb8d6ce6e065b25

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,621 @@ float Lb302FilterIIR2::process(const float& samp)
-		m_notes.prepend( _n );
+		m_trueFreq = nph->frequency();
+		const auto true_inc = phaseInc(m_trueFreq);
+		if (m_slideToggle.value()) { m_slideBase = true_inc; } else { m_vcoInc = true_inc; }
```
```suggestion
		const auto trueInc = phaseInc(m_trueFreq);
		if (m_slideToggle.value()) { m_slideBase = trueInc; } else { m_vcoInc = trueInc; }
```
Naming convention

rubiefawn replied:
Resolved as of 19023f4e9583f6c0455db3c8bcb8d6ce6e065b25

messmerd commented on plugins/Lb302/Lb302.h:
```diff
@@ -71,191 +66,246 @@ class Lb302FilterKnobState
+	sample_t process(sample_t samp) override;
+
+protected:
+	float m_kfcn;
+	float m_kp;
+	float m_kp1h; //!< Cached value of (@ref m_kp + 1) / 2
+	float m_kres;
+	std::array<sample_t, 3> m_ay = { 0.f, 0.f, 0.f };
+	float m_lastin = 0.f;
+	float m_value;
```
I think I'd feel better if all data members were initialized

rubiefawn replied:
- `m_kfcn` is always initialized in all code paths before it is used
- `m_kp` might be used before it is initialized? (I thought the constructor called `envRecalc()` but that doesn't seem to be the case anymore)
- `m_kp1h` has the same issues as `m_kp` as it is dependent on `m_kp`
- `m_kres` has the same issues
- `m_value` has the same issues

rubiefawn replied:
Resolved as of bfc93077536cdd7296487c11376d19239bebf572

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,621 @@ float Lb302FilterIIR2::process(const float& samp)
 	{
-		m_notes.prepend( _n );
+		m_trueFreq = nph->frequency();
+		const auto true_inc = phaseInc(m_trueFreq);
```
```suggestion
		const auto trueInc = phaseInc(m_trueFreq);
```
It looks like `m_trueFreq` isn't initialized upon construction, and I'm not sure if it could be read before being initialized. Should probably set it to something.

rubiefawn replied:
Resolved as of 73b92f4f532b016c81fecc6dd4f5731d201d0520

messmerd commented on plugins/Lb302/Lb302.h:
```diff
@@ -71,191 +66,246 @@ class Lb302FilterKnobState
-	      vca_a0,           // Initial amplifier coefficient
-	      vca_a;            // Amplifier coefficient.
+	f_cnt_t m_vcfEnvPos = s_envInc; //!< Update counter. Updates when >= @ref s_envInc
+	f_cnt_t m_releaseFrame;
```
Looks like this might be read before it is initialized.

rubiefawn replied:
Resolved as of 73b92f4f532b016c81fecc6dd4f5731d201d0520

messmerd commented on plugins/Lb302/Lb302.h:
```diff
@@ -71,191 +66,246 @@ class Lb302FilterKnobState
+	//! adjusted as needed, but it must always be a power of 2.
+	//!
+	//! @see m_notes
+	static constexpr size_t s_maxPendingNotes = 128;
```
I'd prefer using `std::size_t` here and elsewhere, but it's not super important

rubiefawn replied:
Resolved as of 19023f4e9583f6c0455db3c8bcb8d6ce6e065b25

rubiefawn committed:
Resolve some code review comments

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

rubiefawn committed:
Fix more code review comments

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

rubiefawn committed:
Run envRecalc() in Lb302Filter3Pole() constructor

This properly initializes all its member variables that aren't set using
NSDMI

messmerd commented on include/Hardware.h:
```diff
@@ -0,0 +1,151 @@
+		#include <intrin.h>
+	#elif defined(__GNUG__) // HACK: Remove this once GCC properly provides __ARM_ACLE
+		#if defined(LMMS_HOST_ARM64)
+			// https://developer.arm.com/documentation/ddi0602/2026-03/Base-Instructions/ISB--Instruction-synchronization-barrier-
+			// 15 is the only allowed value for the parameter, so just ignore it and use 15 lol
+			inline void __isb(unsigned int) { asm volatile ("isb 15" ::: "memory"); }
```
Is ISB really what we want? It seems unrelated to spin locks.

Why not use the `yield` ASM instruction like you do for 32-bit ARM?

rubiefawn replied:
https://github.com/rust-lang/rust/commit/c064b6560b7ce0adeb9bbf5d7dcf12b1acb0c807

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,627 @@ float Lb302FilterIIR2::process(const float& samp)
+	auto tries = s_maxNoteEnqueueRetries;
+	auto writeClaimedExpected = m_notesWriteClaimed.load(std::memory_order_relaxed);
+	std::size_t index;
+	std::size_t next_index;
```
```suggestion
	std::size_t nextIndex;
```
Missed this one

rubiefawn replied:
Resolved in 37de8f76aa977714117816092ca978838ea1f472

messmerd commented on plugins/Lb302/Lb302.cpp:
What are your thoughts on generalizing the new MPSC note processing you added in this PR to something that works for other instruments as well? Something that could let us hide most of the messy details in `Instrument.cpp` or somewhere else, and provide plugin developers with a nice, easy-to-use note queue which would be passed to `play()` for instruments to read from.

I'm not suggesting any of this should be done in this PR, of course, but it's something that would help with #8275, and note events are one of the few aspects of the new plugin API that I don't have a solid plan for.

rubiefawn replied:
I think that's a good idea. The code for the MPSC note queue is taken from a project of mine: https://codeberg.org/rubiefawn/rtchan

I haven't meaningfully worked on it since I started this PR, but the other queue types can all be derived from MPMC, which is why that's the only one with any real code in it lol.

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,627 @@ float Lb302FilterIIR2::process(const float& samp)
+	sawWaveBtn->move(waveBtnX, waveBtnY);
+	sawWaveBtn->setActiveGraphic(embed::getIconPixmap("saw_wave_active"));
+	sawWaveBtn->setInactiveGraphic(embed::getIconPixmap("saw_wave_inactive"));
+	sawWaveBtn->setToolTip(tr("Click here for a saw-wave."));
```
```suggestion
	sawWaveBtn->setToolTip(tr("Click here for a saw wave."));
```
Maybe we should take the opportunity to remove the weird unnecessary hyphen from some of these tooltips?

rubiefawn replied:
Resolved in 37de8f76aa977714117816092ca978838ea1f472

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,627 @@ float Lb302FilterIIR2::process(const float& samp)
+	triangleWaveBtn->move(waveBtnX + (16 * 1), waveBtnY);
+	triangleWaveBtn->setActiveGraphic(embed::getIconPixmap("triangle_wave_active"));
+	triangleWaveBtn->setInactiveGraphic(embed::getIconPixmap("triangle_wave_inactive"));
+	triangleWaveBtn->setToolTip(tr("Click here for a triangle-wave."));
```
```suggestion
	triangleWaveBtn->setToolTip(tr("Click here for a triangle wave."));
```

rubiefawn replied:
Resolved in 37de8f76aa977714117816092ca978838ea1f472

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,627 @@ float Lb302FilterIIR2::process(const float& samp)
+	sqrWaveBtn->move(waveBtnX + (16 * 2), waveBtnY);
+	sqrWaveBtn->setActiveGraphic(embed::getIconPixmap("square_wave_active"));
+	sqrWaveBtn->setInactiveGraphic(embed::getIconPixmap("square_wave_inactive"));
+	sqrWaveBtn->setToolTip(tr("Click here for a square-wave."));
```
```suggestion
	sqrWaveBtn->setToolTip(tr("Click here for a square wave."));
```

rubiefawn replied:
Resolved in 37de8f76aa977714117816092ca978838ea1f472

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,627 @@ float Lb302FilterIIR2::process(const float& samp)
+	roundSqrWaveBtn->move(waveBtnX + (16 * 3), waveBtnY);
+	roundSqrWaveBtn->setActiveGraphic( embed::getIconPixmap("round_square_wave_active"));
+	roundSqrWaveBtn->setInactiveGraphic( embed::getIconPixmap("round_square_wave_inactive"));
+	roundSqrWaveBtn->setToolTip(tr("Click here for a square-wave with a rounded end."));
```
```suggestion
	roundSqrWaveBtn->setToolTip(tr("Click here for a square wave with a rounded end."));
```

rubiefawn replied:
Resolved in 37de8f76aa977714117816092ca978838ea1f472

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,627 @@ float Lb302FilterIIR2::process(const float& samp)
+	sinWaveBtn->move(waveBtnX + (16 * 5), waveBtnY);
+	sinWaveBtn->setActiveGraphic(embed::getIconPixmap("sin_wave_active"));
+	sinWaveBtn->setInactiveGraphic(embed::getIconPixmap("sin_wave_inactive"));
+	sinWaveBtn->setToolTip(tr("Click for a sine-wave."));
```
```suggestion
	sinWaveBtn->setToolTip(tr("Click for a sine wave."));
```

rubiefawn replied:
Resolved in 37de8f76aa977714117816092ca978838ea1f472

messmerd commented on plugins/Lb302/Lb302.cpp:
```diff
@@ -197,844 +153,627 @@ float Lb302FilterIIR2::process(const float& samp)
+	whiteNoiseWaveBtn->move(waveBtnX + (16 * 7), waveBtnY);
+	whiteNoiseWaveBtn->setActiveGraphic(embed::getIconPixmap("white_noise_wave_active"));
+	whiteNoiseWaveBtn->setInactiveGraphic(embed::getIconPixmap("white_noise_wave_inactive"));
+	whiteNoiseWaveBtn->setToolTip(tr("Click here for white-noise."));
```
```suggestion
	whiteNoiseWaveBtn->setToolTip(tr("Click here for white noise."));
```

rubiefawn replied:
Resolved in 37de8f76aa977714117816092ca978838ea1f472

rubiefawn committed:
Change some tooltips

messmerd reviewed (approved):
LGTM

rubiefawn committed:
Add some more documentation to Hardware.h

rubiefawn commented:
Class member layout could probably be optimized for better locality, but feature freeze is in like two days lol. That can be tidied up after the freeze is over. Merging  immediately !!!🥝!!!
